machine_learning_hep/globalfitter.py

```diff
@@ -58,255 +28,280 @@ def fixpar(massmin, massmax, masspeak, range_signal):
-    par_fix1, par_fix2, par_fix3, par_fix4, par_fix5 =\
+
+def update_check_signal_fit(sig_fit):
+    error_list = []
+    int_sig = sig_fit.GetParameter(0)
+    mean = sig_fit.GetParameter(1)
+    sigma = sig_fit.GetParameter(2)
+    if int_sig < 0. < sigma or sigma < 0. < int_sig:
+        error_list.append(f"Both integral pre-factor and sigma have to have the same sign. " \
+                          f"However, pre-factor is {int_sig} and sigma is {sigma}.")
+    if mean < 0.:
+        error_list.append(f"Mean is negative: {mean}")
+
+    if error_list:
+        return "\n".join(error_list)
+
+    # Seems sane, set both sigma and int_sig to positive values
+    sig_fit.SetParameter(0, abs(int_sig))
+    sig_fit.SetParameter(2, abs(sigma))
+
+    return ""
+
+def signal_func(sgnfunc):
+    if sgnfunc != "kGaus":
+        get_logger().fatal("Unknown signal fit function %s", sgnfunc)
+    return "[0]/(sqrt(2.*pi))/[2]*(exp(-(x-[1])*(x-[1])/2./[2]/[2]))"
+
+def pol1_func_sidebands(xval, par):
+    if abs(xval[0] - par[4]) < par[5]:
+        TF1.RejectPoint()
+    return par[0] / par[2] + par[1] * (xval[0] - 0.5 * par[3])
+
+def pol2_func_sidebands(xval, par):
+    if abs(xval[0] - par[6]) < par[7]:
+        TF1.RejectPoint()
+    return par[0] / par[3] + par[1] * (xval[0] - 0.5 * par[4]) + par[2] * \
+           (xval[0] * xval[0] - 1/3. * par[5] / par[3])
+
+def bkg_fit_func(func_name, func_type, massmin, massmax, integralhisto, masspeak, range_signal):
+    # Immediately exit if function is unknown
+    if func_type not in ["Pol1", "Pol2"]:
+        get_logger().fatal("Unkown background fit dunction %s", func_type)
+
+    par_fix1, par_fix2, par_fix3, par_fix4, par_fix5 = \
         fixpar(massmin, massmax, masspeak, range_signal)
-    if fitsidebands:
-        if bkgfunc == "Pol1":
-            back_fit_new = TF1("back_fit_new", pol1_func_sidebands, massmin, massmax, 6)
-            back_fit_new.SetParNames("BkgInt", "Slope", "", "", "", "")
-            back_fit_new.SetParameters(integralhisto, -100.)
-            back_fit_new.FixParameter(2, par_fix1)
-            back_fit_new.FixParameter(3, par_fix2)
-            back_fit_new.FixParameter(4, par_fix4)
-            back_fit_new.FixParameter(5, par_fix5)
-        if bkgfunc == "Pol2":
-            back_fit_new = TF1("back_fit_new", pol2_func_sidebands, massmin, massmax, 8)
-            back_fit_new.SetParNames("BkgInt", "Coef1", "Coef2", "", "", "", "", "")
-            back_fit_new.SetParameters(integralhisto, -10., 5)
-            back_fit_new.FixParameter(3, par_fix1)
-            back_fit_new.FixParameter(4, par_fix2)
-            back_fit_new.FixParameter(5, par_fix3)
-            back_fit_new.FixParameter(6, par_fix4)
-            back_fit_new.FixParameter(7, par_fix5)
-        return back_fit_new
-    else:
-        if bkgfunc == "Pol1":
-            back_refit_new = TF1("back_refit", pol1_func, massmin, massmax, 4)
```

Comment:
The *stripped* versions of the fit functions (here "back_fit_new" vs. "back_refit") are not necessary. There can be multiple instances of the same function (of course with different names) such that parameters can be set and we have different objects `back_fit` and `back_refit` in the code below. **NOTE** that the latter one is only used to store parameters as it is also the case for `sig_fit`.
Only the object `back_fit` and `tot_fit` are actually fitted.